### ISSUE TITLE
+= on array causes java.lang.RuntimeException

### ISSUE THREAD

**Issue by chosia** [google/closure-compiler#4014]
```
// Show closure-compiler throw a java.lang.RuntimeException.

// Compile with "closure-compiler bug423.js" will give:
//    Invalid child for ASSIGN_ADD node.

// Compile with "closure-compiler -O ADVANCED bug423.js" will give:
//    INTERNAL COMPILER ERROR. ...
//    Type inference doesn't know to handle token ARRAY_PATTERN

let N = 3

// This is ok.

for (let [i,j] = [N-1,0]; j < N; [i,j]   =  [j, j+1]) {
    console.log(">>>", i,"-",j)
}

// This makes the closure-compiler crash. Note the ''+=''.

for (let [i,j] = [N-1,0]; j < N; [i,j]  +=  [j, j+1]) {
    console.log(">>>", i,"-",j)
}
```

**Comment by lvelden** [comment]
It looks like the respective cases are not implemented, see https://github.com/google/closure-compiler/blob/d3337949f6d1283fffd602b0f19e6204b9d0168a/src/com/google/javascript/jscomp/AstValidator.java#L1734 and https://github.com/google/closure-compiler/blob/d3337949f6d1283fffd602b0f19e6204b9d0168a/src/com/google/javascript/jscomp/TypeInference.java#L855

Also, for reference , the AST of the stack trace for `$ google-closure-compiler bug423.js`:

```
java.lang.IllegalStateException: Invalid child for ASSIGN_ADD node. Reference node:
ARRAY_PATTERN 3:33  [length: 5] [source_file: bug423.js]
    NAME b 3:34  [length: 1] [source_file: bug423.js]
    NAME a 3:36  [length: 1] [source_file: bug423.js]
 
 Parent node:
ASSIGN_ADD 3:33  [length: 19] [source_file: bug423.js]
    ARRAY_PATTERN 3:33  [length: 5] [source_file: bug423.js]
        NAME b 3:34  [length: 1] [source_file: bug423.js]
        NAME a 3:36  [length: 1] [source_file: bug423.js]
    ARRAYLIT 3:44  [length: 8] [source_file: bug423.js]
        NAME a 3:45  [length: 1] [source_file: bug423.js]
        ADD 3:48  [length: 3] [source_file: bug423.js]
            NAME a 3:48  [length: 1] [source_file: bug423.js]
            NUMBER 1.0 3:50  [length: 1] [source_file: bug423.js]
```

**Comment by rishipal** [comment]
The code snippet should produce a closure compiler parse error instead of a runtime crash.  

On chrome, this gives:
```
Uncaught SyntaxError: Invalid left-hand side in assignment
```

**Comment by rishipal** [comment]
Fixed with c8579f86232ef872d01df6e2f99a00757f993f67
